include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -132,15 +140,74 @@ struct _HasDefaultName<_ExternalName<_InternalName>>
     static constexpr bool value = _HasDefaultName<_InternalName>::value;
 };
 
+template <typename... _Name>
+struct __optional_kernel_name;
+
+template <typename _CustomName>
+using __kernel_name_provider =
+#if __SYCL_UNNAMED_LAMBDA__
+    typename ::std::conditional<_HasDefaultName<_CustomName>::value, __optional_kernel_name<>,
+                                __optional_kernel_name<_CustomName>>::type;
+#else
+    __optional_kernel_name<_CustomName>;
+#endif
+
+template <char...>
+struct __composite_kernel_name
+{
+};
+
+#if __has_builtin(__builtin_sycl_unique_stable_name)
+template <typename _Tp>
+class __kernel_name_composer
+{
+    static constexpr auto __name = __builtin_sycl_unique_stable_name(_Tp);
+    static constexpr ::std::size_t __name_size = __builtin_strlen(__name);
+
+    template <::std::size_t... _Is>
+    static __composite_kernel_name<__name[_Is]...>
+    __compose_kernel_name(oneapi::dpl::__internal::__index_sequence<_Is...>);
+
+  public:
+    using type = decltype(__compose_kernel_name(oneapi::dpl::__internal::__make_index_sequence<__name_size>{}));
+};
+#endif // __has_builtin(__builtin_sycl_unique_stable_name)
+
 template <template <typename...> class _BaseName, typename _CustomName, typename... _Args>
 using _KernelName_t =
```

Comment:
Now we should use ```_KernelName_t``` for kernels which may require precompilation and ```__kernel_name_provider``` otherwise. Let's show this in their names or make a comment at least. We can also make a single entity for that with e.g. property which shows if the precompilation is necessary.

Reply:
I agree with your concern. Please suggest the names

Reply:
_kernel_name and _precompiled_kernel_name?

Reply:
Created the issues. I believe I have not so bad ideas